### ISSUE TITLE
LSP SemanticTokens offset incorrect in some cases

### ISSUE THREAD

**Issue by jnt0r** [eclipse/xtext#2715]
The offset in SemanticTokens response is slightly off in some cases. I debugged this and found out, that the wrong offset is calculated deeper in xtext. The TextRegion is already filled with the wrong offset, which is then used to calculate the line and character in the SemanticTokensService. I don't know and yet didn't find where the TextRegion is created and filled with the offset.

It looks like line ending characters (\n) are counted in the offset as well:

![grafik](https://github.com/eclipse/xtext/assets/38696668/abda86f9-2f62-4d9f-a675-9df87e3fc18a)

![grafik](https://github.com/eclipse/xtext/assets/38696668/7c4234e2-81bf-4754-96ee-508bcdbb19b9)

![grafik](https://github.com/eclipse/xtext/assets/38696668/27eadc16-e812-4e6c-855f-0d070dba3be7)

**Comment by cdietrich** [comment]
Xtext offsets are for the complete document and this include whitespace 

so the question is if they are correctly translated to line and column for lsp 

a sample grammar and unit test might be helpful

**Comment by jnt0r** [comment]
I try to take a look into this in the next days.

**Comment by jnt0r** [comment]
I think adjusting the way of calculating offsets is very complicated, as it is used widely in xtext and lays deeply in the core. Therefore, I am currently taking a look at the PositionReader that is returning an LSP4j Position based on an InputStream and an offset. I try to adjust it, so it ignores newlines.

**Comment by jnt0r** [comment]
Update: The Problem lays in the underlying LineNumberReader which is used in the PositionReader. The LineNumberReader skips \r characters and does not count them as a character when reading.

![grafik](https://github.com/eclipse/xtext/assets/38696668/8a7b9e5f-1bc6-43d0-b258-3aab60434cde)

When the offset is calculated, it includes \r characters. Therefore, we get a wrong position, as these characters are skipped when resolving the position to this offset. 

As the LineNumberReader is part of the java.io package, I try to work around this behavior in the PositionReader by detecting if carriage return characters are used and then subtracting the line number from the calculated column. Still not 100% happy with my solution, so will think about a bit more.

**Comment by szarnekow** [comment]
Can you provide an isolated test case that illustrates what you are observing? 
Something like 
```
var reader = new PositionReader("some string literal");
reader.read(..)
...
assertEquals(.. , reader.getPosiiton())
```
?

**Comment by jnt0r** [comment]
I created this branch in my fork: https://github.com/jnt0r/xtext/tree/positionreader-demo-test with this test https://github.com/eclipse/xtext/compare/main...jnt0r:xtext:positionreader-demo-test

I'm not sure if we can assume that all line breaks contain carriage returns, or if there can be a mix of single \n and  \r\n.
Normally, this should be specific to the OS and therefore consistent unless somebody manipulates the file.
My TestCase currently contains a mix of \n and \r\n.

**Comment by jnt0r** [comment]
I just created a PR with my current solution.

**Comment by jnt0r** [comment]
I closed my PR in favor of https://github.com/eclipse/xtext/pull/2735

**Comment by jnt0r** [comment]
Fixed by #2735
